### ISSUE TITLE
`@dynamic` is suppressing a metadata error

### ISSUE THREAD

**Issue by alan-agius4** [angular/angular#21123]
## I'm submitting a...

<pre><code>
[] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
I am bundling a library and having set `skipTemplateCodegen` and `strictMetadataEmit` to `true`.

When passing a lambda funtion to a `decorator` an metadata error is emitted during compilation. Marking it with `@dynamic` has no effect, and in fact I get 
```
Error encountered in metadata generated for exported symbol 'Foo': Metadata collected contains an error that will be reported at runtime: Lambda not supported.
```

When I switch `skipTemplateCodegen` and `strictMetadataEmit` to `false` I also have an error 
```
TypeError: Cannot read property 'kind' of undefined
```

## Expected behavior
Lambda calls can be used in decorators with the `@dynamic`. In all fairness I am still a bit confused for what to use the `@dynamic`, And my question would be is there a way to use lamba function inside decorators or this is by design that you cannot use it?

## Minimal reproduction of the problem with instructions
https://github.com/alan-agius4/angular5-dynamic-supress-meta-error

## What is the motivation / use case for changing the behavior?
In Angular 4 this used to work because the error was not emitted as I believe there was no validation happening, I checked the metadata and indeed there is an error in the meta data.
```
{
	"__symbolic": "method",
	"decorators": [
		{
			"__symbolic": "error",
			"message": "Function call not supported",
			"line": 7,
			"character": 6,
			"module": "./src/foo"
		}
	]
}
```

## Environment

<pre><code>
Angular version: 5.1.X

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.0
- Platform:  Windows
</code></pre>

Others:
Looping in @chuckjaz as this is following #21111

**Comment by alan-agius4** [comment]
From my debugging it seems that  `node.getStart()`
in
https://github.com/angular/angular/blob/f841fbe60fe1db62f700d03d4e084d864e9ab952/packages/compiler-cli/src/metadata/collector.ts#L673

is not returning the correct value

**Comment by alan-agius4** [comment]
@chuckjaz I managed to get this working however the `@dynamic` had to be done as follows

```ts
export function foo(_test: any) {
  return (_target: any, _name: string, _descriptor: PropertyDescriptor) => { };
}

// @dynamic
export class Foo {

  @foo(
    () => ""
  )
  bar() {
    // stub
  }

}
```

**Comment by alan-agius4** [comment]
I managed to get this working, but there are a couple of catches. 

1.  I had to put the `// @dynamic` on the class rather than the exported function. 
2. When doing so I had a TypeScript error `TypeError: Cannot read property 'kind' of undefined`
3. Update to `TypeScript@next` aka `TypeScript 2.7`

```ts
export function foo(_test: any) {
  return (_target: any, _name: string, _descriptor: PropertyDescriptor) => { };
}

// @dynamic
export class Foo {

  @foo(
    () => ""
  )
  bar() {
    // stub
  }

}
```

I am not sure if this is how it should be, mainly because of lack of documentation about the `dynamic` and where it should be placed, but checking the `collector` in the compiler it seems that it's good to place it there.

The method that seems to be working good with TS 2.7 which is not with other TS versions is;
https://github.com/angular/angular/blob/f841fbe60fe1db62f700d03d4e084d864e9ab952/packages/compiler-cli/src/metadata/collector.ts#L671

**Comment by chuckjaz** [comment]
The comment should have worked.

Partial for documentation for this is coming in #21135.

**Comment by alan-agius4** [comment]
Relates to https://github.com/Microsoft/TypeScript/issues/17384

Also, I from debugging a bit I think that the `@dynamic` should be on the actual class for 2 reasons;
1) The metadata error is related to the class `Error encountered in metadata generated for exported symbol `Foo`
2) If the decorator implementation is in a separate file it won't work as the compiler is using using AST for the @dynamic but rather the file source code that the metadata error is in.

So I believe this is working as intended if you take out the fact that `TypeScript` gives out an error if you put the `// @dynamic` comment on top of the class, but that should is solved in `TS 2.7` which is a bit of a bummer.

**Comment by JoostK** [comment]
Closing as obsolete given that the metadata system is being phased out.
